proxy: Make NACK cancel the WaitGroup and pass NACK detail in ProxyError

jrajahalme opened the pull request:
Make received NACKs cancel the WaitGroup immediately.

Pass NACK error detail in a new error type (ProxyError).

Change xDS versioning start at 1, so that we can map the initial (no-)version ("") from Envoy to 0.

Fixes: #5649 
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4535)
<!-- Reviewable:end -->

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -103,7 +99,9 @@ type Completion struct {
-	callback func()
+	// 'success' tells if the operation was successful or not.
+	// returns updated version number or 0.
+	callback func(success bool) uint64
```
This signature is not OK, it's breaking the genericity of this data structure.
I'm ok with passing an argument to the callback. But there's no need to constrain it to a `bool`, so replace `success bool` with `arg interface{}`.

And returning an `uint64` is too specific to your use case. That's a no-no.
I think you can design this without adding this return value.

jrajahalme replied:
I understand that it is easy for the callback to interpret a generic arguments as it needs to, but the call sites still need to know what the callbacks can accept, no?

Maybe it would be better to add an optional separate callback for the failed completion? Then I would not need to touch the existing callback implementations.

The design intent here is to allow the original completion keep waiting through optional retries. For that to work the version number that the pending completion is waiting for needs to be updated to match the modified version of the resource. Due to the callbacks being called while `AckingResourceMutatorWrapper`'s lock is being held, the callbacks can not call back to the `AckingResourceMutatorWrapper` to Upsert a new version of the resource. In this version of the PR I opted to the owner of the resource (XDSServer) modify the resource itself, without going through `AckingResouerceMutatorWrapper`, and then return the version number of the updated resource. Version numbers are meaningful for the users of completions, but I see that Completion is separated into a package and it is not aware of versions. One alternative would be to launch a goroutine from the NACK callback and call to `AckingResourceMutatorWrapper` from there, and make `AckingResourceMutatorWrapper.Upsert()` check if the passed Completion is already in the list of pending completions. If yes, then update it instead of creating a new pending completion with the same `completion.Completion`. Would this be acceptable?

rlenglet reviewed (changes requested):
Only looked at the changes to the Completion type so far.

rlenglet commented on pkg/completion/completion_test.go:
```diff
@@ -190,6 +190,40 @@ func (s *CompletionSuite) TestCompletionWithCallback(c *C) {
+	comp.Complete()
+	comp.Complete()
+	comp.Complete()
+
```
I would also call `comp.Retry()` here again, to show that it has no effect after completion.

jrajahalme replied:
OK, somehow I thought I did that already...

jrajahalme replied:
Added

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -237,7 +237,25 @@ func (s *XDSServer) AddListener(name string, endpointPolicyName string, port uin
+			go func() {
+				// Upsert a new version of the Listener
+				// TODO: Try to recover by reallocating the proxy port
+				s.listenerMutator.Upsert(ListenerTypeURL, name, listenerConf, []string{"127.0.0.1"}, comp)
```
Have you tested that this works?
If I read this correctly, `comp` is `nil` by the time this closure is created. This is definitely not what you want.
I think you need to create a new `Completion` here.

You have a circular dependency problem.
You can't create the completion without the closure, and you can't create the closure without the completion.
:/

jrajahalme replied:
Not tested yet, but closure should not cache variable values. `comp` is a pointer and it is set to non-nil value before the closure is used. At least in the tests incrementing a counter in the completion callback works.

jrajahalme replied:
I added a server_e2e_test with this same pattern and it works!

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -112,19 +118,50 @@ func (m *AckingResourceMutatorWrapper) Upsert(typeURL string, resourceName strin
-		remainingNodesResources: make(map[string]map[string]struct{}, len(nodeIDs)),
-	}
+	comp, found := m.pendingCompletions[completion]
+	if found {
```
Have you checked that this branch of the code is hit at all?

jrajahalme replied:
New unit tests should cause this to be hit. I.e., original Completion completes successfully after a retry.

jrajahalme replied:
Added a server_e2e_test that exercises this. Added a log line and verified that it is seen when the test is run.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -112,19 +118,50 @@ func (m *AckingResourceMutatorWrapper) Upsert(typeURL string, resourceName strin
+	} else {
+		// Create a new pendingCompletion
+		comp = &pendingCompletion{
+			completion:              completion,
```
If you move that into the key of your new map, remove it from the `pendingCompletion` struct altogether.

jrajahalme replied:
Right, will do!

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -210,7 +253,53 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
+	if err != nil {
+		// Ignore NACKs from unknown or misconfigured nodes which have invalid
+		// node identifiers.
+		ackLog.Warning("invalid ID in Node identifier; ignoring NACK")
```
invalid -> Invalid

jrajahalme replied:
Fixed.

jrajahalme replied:
There were two of these, will fix this one too.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -210,7 +253,53 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
+
+	remainingCompletions := make(map[*completion.Completion]*pendingCompletion, len(m.pendingCompletions))
+
+	for _, comp := range m.pendingCompletions {
```
You can now use the key of that:
`for comp, pending := range m.pendingCompletions {`

jrajahalme replied:
Done.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -210,7 +253,53 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
+	m.pendingCompletions = remainingCompletions
+}
+
+func (m *AckingResourceMutatorWrapper) HandleResourceVersionNack(version uint64, node *envoy_api_v2_core.Node, resourceNames []string, typeURL string) {
```
Add a comment.

jrajahalme replied:
Added.

rlenglet reviewed (changes requested):
I really like the new Retry design!
I want to make sure that the usage of `wg.AddCompletionWithCallbacks` is correct in `server.go`, i.e. that the `comp` value in the closure is not `nil`.

rlenglet reviewed (approved):
Some of my comments still need to be addressed, but this looks good overall!

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -215,7 +214,18 @@ retryCreatePort:
+				newPort, err := p.allocatePort()
+				if err == nil {
+					delete(p.allocatedPorts, redir.ProxyPort)
+					redir.ProxyPort = newPort
```
You're not holding any mutex when modifying `redir`? That's probably not thread-safe. There's a good chance this update won't be seen by other goroutines.

jrajahalme replied:
Added proper locking as a separate commit so that it is easier to review.

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -215,7 +214,18 @@ retryCreatePort:
+				// Reallocate proxyport after a NACK has been received.
+				p.mutex.Lock()
+				newPort, err := p.allocatePort()
+				if err == nil {
```
You must return `err` if not `nil`. Right now, in case of error, you return an unspecified `newPort` value and a `nil` error.

jrajahalme replied:
Fixed, thanks for spotting this!

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -71,8 +71,7 @@ type Proxy struct {
 
-	// allocatedPorts is a map of all allocated proxy ports pointing
-	// to the redirect rules attached to that port
+	// allocatedPorts is a map of all allocated proxy ports
```
`a map` -> `the map`

jrajahalme replied:
Fixed.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -238,7 +238,33 @@ func (s *XDSServer) AddListener(name string, endpointPolicyName string, port uin
+		if err != nil && retries < maxRetries {
+			retries++
+			listenerConf.Address.GetSocketAddress().PortSpecifier = &envoy_api_v2_core.SocketAddress_PortValue{PortValue: uint32(port)}
+			log.Debugf("Retrying listener %s with reallocated proxyport (%u -> %u), completion %v", name, oldPort, port, comp)
```
Is the Completion readable at all in the logs? Is it really useful to log?

jrajahalme replied:
This was a remnant from debugging, wanted to see it is not `nil`, I'll remove it.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -238,7 +238,33 @@ func (s *XDSServer) AddListener(name string, endpointPolicyName string, port uin
+				log.Debug("Envoy: Listener updated after NACK")
+			}()
+		} else {
+			log.WithError(err).Errorf("Envoy: Failed to apply new listener configuration after %u retries (irrecoverable NACK received), removing listener %s", retries, name)
```
`%u` -> `%d`

jrajahalme replied:
fixed.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -238,7 +238,33 @@ func (s *XDSServer) AddListener(name string, endpointPolicyName string, port uin
+		if err != nil && retries < maxRetries {
+			retries++
+			listenerConf.Address.GetSocketAddress().PortSpecifier = &envoy_api_v2_core.SocketAddress_PortValue{PortValue: uint32(port)}
+			log.Debugf("Retrying listener %s with reallocated proxyport (%u -> %u), completion %v", name, oldPort, port, comp)
```
`%u` -> `%d`

jrajahalme replied:
Fixed.

rlenglet commented on pkg/proxy/kafka.go:
```diff
@@ -82,6 +82,7 @@ func createKafkaRedirect(r *Redirect, conf kafkaConfiguration, endpointInfoRegis
 
 	// Listen needs to be in the synchronous part of this function to ensure that
 	// the proxy port is never refusing connections.
+	// redirect not visible yet, no need to lock for ProxyPort
```
Remove those new comments in kafka.go about locking for accesses to ProxyPort.
The redirect is never modified in the case of Kafka.

rlenglet commented on pkg/proxy/redirect.go:
```diff
@@ -75,7 +75,9 @@ func (r *Redirect) updateRules(l4 *policy.L4Filter) {
 // and will remove the proxymap entry for that connection
 func (r *Redirect) removeProxyMapEntryOnClose(c net.Conn) error {
+	r.mutex.RLock()
 	key, err := getProxyMapKey(c, r.ProxyPort)
```
I would only lock the lookup of `r.ProxyPort`:
```
r.mutex.RLock()
port := r.ProxyPort
r.mutex.RUnlock()
key, err := getProxyMapKey(c, port)
...
```

jrajahalme replied:
Done.

rlenglet reviewed (dismissed):
This is mostly LGTM.
But it's still incomplete.
The port that was allocated in the first try is returned by `Daemon`'s `UpdateProxyRedirect` method, and stored in `Endpoint`'s `desiredRedirects` and `realizedRedirects` maps.
When you reallocate a redirect's port, you must modify those maps and regenerate the BPF program whenever you do so (the redirect ports are currently compiled into BPF). My understanding is that your implementation here doesn't do that.

eloycoto commented on pkg/envoy/server.go:
```diff
@@ -238,7 +238,33 @@ func (s *XDSServer) AddListener(name string, endpointPolicyName string, port uin
+				// Upsert a new version of the Listener
+				// This will block until the caller has released locks.
+				s.listenerMutator.Upsert(ListenerTypeURL, name, listenerConf, []string{"127.0.0.1"}, comp)
+				log.Debug("Envoy: Listener updated after NACK")
```
Could we add the log of the port? It would be useful to debug issues.

eloycoto commented on pkg/envoy/server.go:
```diff
@@ -238,7 +238,33 @@ func (s *XDSServer) AddListener(name string, endpointPolicyName string, port uin
+				log.Debug("Envoy: Listener updated after NACK")
+			}()
+		} else {
+			log.WithError(err).Errorf("Envoy: Failed to apply new listener configuration after %d retries (irrecoverable NACK received), removing listener %s", retries, name)
```
And here if you added the error it'll be nice too.

eloycoto commented on pkg/envoy/xds/server_e2e_test.go:
```diff
@@ -1125,3 +1126,299 @@ func (s *ServerSuite) TestRequestHighVersionFromTheStart(c *C) {
 	}
 }
+
+func (s *ServerSuite) TestNAckWithRetry(c *C) {
```
Love this one :-)

rlenglet commented on daemon/daemon.go:
```diff
@@ -155,18 +155,20 @@ type Daemon struct {
+// endpoint using the provided L4 filter. Returns the allocated proxy port via
+// 'acked' callback, which will be called if the operation is successful
+// before the 'proxyWaitGroup' is canceled.
+func (d *Daemon) UpdateProxyRedirect(e *endpoint.Endpoint, l4 *policy.L4Filter, proxyWaitGroup *completion.WaitGroup, acked func(redirectPort uint16), nacked func(error)) error {
```
This definitely conflicts with the changes I made in #5451 to finalize (resp. revert) xDS changes on regeneration success (resp. failure).
I would favor removing your changes here to the signature of `UpdateProxyRedirect` and other related methods, because the finalize/revert approach in #5451 is more general as it also covers NACKs on network policy updates and redirect removals.

jrajahalme replied:
Reverted these changes.

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -67,33 +68,47 @@ func (wg *WaitGroup) AddCompletion() *Completion {
+// No callbacks of the completions in this wait group will be called after
+// this returns.
+// Returns the error value of one of the completions, if available, or the
+// error value of the WaitGroup otherwise.
```
WaitGroup -> Context (?)

jrajahalme replied:
right.

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -67,33 +68,47 @@ func (wg *WaitGroup) AddCompletion() *Completion {
 	for i, comp := range wg.pendingCompletions {
 		select {
 		case <-comp.Completed():
+			err = comp.Error
```
You shouldn't overwrite any previous error if `comp.Error` is `nil`.
Also, to be consistent with the other cases below, I would only set `err` if it is currently `nil`.
```
if err == nil && comp.Error != nil {
    err = comp.Error
}
```

jrajahalme replied:
changed

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -67,33 +68,47 @@ func (wg *WaitGroup) AddCompletion() *Completion {
-	// ctx is the context of the wait group.
-	ctx context.Context
+	// wg is the wait group the completion belongs to
+	wg *WaitGroup
```
I would prefer to avoid the circular dependency. Instead of replacing `ctx` with `wg`, I would just add the `cancel` function, since accessing `cancel` is the only reason you did this change.

jrajahalme replied:
This change requires the Completion.Context() method to be removed. Doing that.

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -103,38 +118,43 @@ type Completion struct {
+	callback func(err error)
+
+	// err is the error the completion completed with
+	Error error
```
I would not export this field, since all accesses to this field must be done while locking using `lock`. So I would rename this `Error` -> `Err`, and define a new `Error()` method that returns this field while holding `lock`.

jrajahalme replied:
This field can be ready without locking after calling Complete(), or after the Completed() channel is closed. Added a comment to that effect.

rlenglet replied:
No. It typically can't be read without locking if it's written by another goroutine. Cf. https://golang.org/ref/mem#tmp_2
> When multiple goroutines access a shared variable v, they must use synchronization events to establish happens-before conditions that ensure reads observe the desired writes.

jrajahalme replied:
Made the member non-exported by renaming to `err`. Added a new `Err()` method mirroring the method of the same name of `Context`.

jrajahalme replied:
Added a commit to make it clear and consistent how WaitGroup's Wait() prioritizes error values for return. Also addressed the concern of accessing completion's 'err' without locking.

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -103,38 +118,43 @@ type Completion struct {
+			// that they are actually called before Wait() returns.
+			c.callback(err)
+		}
+		// Terminate the WaitGroup on failure
```
```
// Cancel the context on failure.
```

jrajahalme replied:
changed.

rlenglet reviewed (changes requested):
I've not finished reviewing everything yet.
The completion package changes look good to me, except for a few minor changes. The idea of passing an `error` to `Complete` and canceling the context on failure are great.

rlenglet commented on pkg/endpoint/bpf.go:
```diff
@@ -246,53 +247,124 @@ func hashHeaderfile(hashWriter hash.Hash, filepath string) (hash.Hash, error) {
 	return hashWriter, nil
 }
 
+// AckedRedirect is a map value type that contains minimal state that
```
Cf. my comments on `daemon/daemon.go`, I would revert all your changes to `pkg/endpoint/bpf.go`. They are redundant with and less general than my implementation in #5451.

jrajahalme replied:
right, reverting these changes.

rlenglet commented on pkg/envoy/envoy_test.go:
```diff
@@ -131,8 +137,68 @@ func (s *EnvoySuite) TestEnvoy(c *C) {
+	c.Assert(envoyProxy, NotNil)
+	log.Debug("started Envoy")
+
+	ProxyPort := uint16(80) // Port 80 requires privileges we dont have when running this test
```
That's not true. Unit tests are run with `sudo`, cf. `Makefile`:
```
unit-tests: start-kvstores
    ...
    sudo go test ...
```
You should pick a port that should always be already open, e.g. port `22` (ssh).

jrajahalme replied:
Did not face this as the tests requiring installed `cilium-envoy`are disabled by default. Changing the port to 22, will test sudoed as well.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -24,6 +27,19 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+type ProxyError struct {
```
Add a comment.

jrajahalme replied:
Added.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -24,6 +27,19 @@ import (
+}
+
+var (
+	NackReceived error = errors.New("NACK received")
```
Rename into `ErrNackReceived`.

jrajahalme replied:
Changed.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -93,23 +111,38 @@ type pendingCompletion struct {
+	if completion == nil {
+		log.WithFields(logrus.Fields{
+			logfields.XDSTypeURL: typeURL,
+		}).Debug("No completion given for resource ", resourceName)
```
```
log.WithFields(logrus.Fields{
    logfields.XDSTypeURL: typeURL,
    logfields.XDSResourceName: resourceName,
}).Debug("no completion given to upsert xDS resource")
```

jrajahalme replied:
changed. I also made this a Fatal case.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -93,23 +111,38 @@ type pendingCompletion struct {
+	if found {
+		log.WithFields(logrus.Fields{
+			logfields.XDSTypeURL: typeURL,
+		}).Fatalf("Completion reused for Upsert: %v", completion)
```
```
log.WithFields(logrus.Fields{
    logfields.XDSTypeURL: typeURL,
    logfields.XDSResourceName: resourceName,
}).Fatalf("attempt to reuse completion to upsert xDS resource: %v", completion)
```

jrajahalme replied:
fixed.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -93,23 +111,38 @@ type pendingCompletion struct {
-	// the resource is actually upserted.
+	// Caller is expected to use the completion callbacks to get notified when the resource has been acknowledged
+	// If the context is cancelled no such acknowledgement is made
+	if completion.Context().Err() != nil {
```
You must do that only _after_ testing that `completion` is not `nil`, which is currently done below.

jrajahalme replied:
Yep, noticed that when editing for the other comments :-)

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -137,8 +169,20 @@ func (m *AckingResourceMutatorWrapper) Delete(typeURL string, resourceName strin
-	comp := &pendingCompletion{
-		completion:              completion,
+	if completion == nil {
+		log.WithFields(logrus.Fields{
```
```
log.WithFields(logrus.Fields{
    logfields.XDSTypeURL: typeURL,
    logfields.XDSResourceName: resourceName,
}).Debug("no completion given to delete xDS resource")
```

jrajahalme replied:
fixed.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -137,8 +169,20 @@ func (m *AckingResourceMutatorWrapper) Delete(typeURL string, resourceName strin
+
+	comp, found := m.pendingCompletions[completion]
+	if found {
+		log.WithFields(logrus.Fields{
```
```
log.WithFields(logrus.Fields{
    logfields.XDSTypeURL: typeURL,
    logfields.XDSResourceName: resourceName,
}).Fatalf("attempt to reuse completion to delete xDS resource: %v", completion)
```

jrajahalme replied:
Thanks for the detailed review, fixed.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -202,7 +246,55 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
+	if err != nil {
+		// Ignore NACKs from unknown or misconfigured nodes which have invalid
+		// node identifiers.
+		ackLog.Warning("Invalid ID in Node identifier; ignoring NACK")
```
```
ackLog.WithError(err).Warning(...)
```

jrajahalme replied:
fixed.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -202,7 +246,55 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
+			// Completion was canceled or timed out.
+			// Remove from pending list.
+			ackLog.Debugf("completion context was canceled: %v", pending)
+			comp.Complete(context.Canceled)
```
To be consistent with what you do in the wait group:
```
comp.Complete(comp.Context().Err())
```

jrajahalme replied:
With the changes due to this review the Complete() call is redundant and hence removed.

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -202,7 +246,55 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
+			_, found := pending.remainingNodesResources[nodeID]
+			if found {
+				ackLog.Debugf("Notifying resource owner of NACK: %v", pending)
+				comp.Complete(&ProxyError{Err: NackReceived, Detail: detail})
```
Awesome!

jrajahalme replied:
:-)

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -202,7 +246,55 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
+		}
+
+		// Check if comp's version was NACKed
+		if pending.version == version && pending.typeURL == typeURL {
```
I think it's incorrect to do an `==` test here. If a later version is NACKed, you may never notify a completion's ACK or NACK.
I think you should test `pending.version <= version` like for an ACK. I.e. if a version is NACKed, all previous pending versions are NACKed too.

rlenglet replied:
Actually, can you confirm what is the `version` in a NACK? Is it the version of the last version ACKed? If that's the case, shouldn't we ACK all versions <=version?

jrajahalme replied:
Digging into this revealed a hole I've been digging myself out of the whole day..
To reliably know what was NACKed I had to change Nonce in a response to be the same as VersionInfo. Then in a forthcoming request:

- VersionInfo indicated the last ACKed version
- ResponseNonce, if > VersionInfo, indicates that versions after VersionInfo, up to and including ResponseNonce, were NACKed.

rlenglet reviewed (changes requested):
Finished reviewing the rest of the PR.

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -276,6 +277,15 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
+				var err error
+				nonce, err = strconv.ParseUint(req.ResponseNonce, 10, 64)
+				if err != nil {
+					requestLog.Errorf("invalid response nonce info in xDS request, not a uint64")
```
`Errorf` -> `Error`

jrajahalme replied:
fixed.

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -383,19 +386,20 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
 			}
 
-			state.nonce = nonce
+			state.nonce = resp.Version
```
It seems like `state.nonce` is not used at all anymore. Just remove that field.

jrajahalme replied:
Removed.

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -168,7 +172,7 @@ type perTypeStreamState struct {
 	// nonce is the nonce sent in the last response to a request for this
 	// resource type.
-	nonce string
+	nonce uint64
```
That field is updated, but never read anymore. Just remove it and the code that updates it.

jrajahalme replied:
Done.

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -383,19 +386,20 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
 				Canary:      resp.Canary,
 				TypeUrl:     state.typeURL,
-				Nonce:       nonce,
+				Nonce:       versionStr,
```
The changes re: the nonce are subtle but important. Please add a comment about it in this commit's message.
Also, in the same commit's message, remove this paragraph, which is not true after your more recent changes:
```
We used to consider a redirect as "realized" after we had asked it to
be created. This commit makes a redirect to be considered "realized"
only after receiving a positive acknowledgement from the proxy.
```

jrajahalme replied:
DOne.

rlenglet reviewed (approved):
Awesome! I just have a few minor comments left. Feel free to merge after addressing them.

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -103,39 +134,48 @@ type Completion struct {
-func (c *Completion) Context() context.Context {
-	return c.ctx
+	// err is the error the completion completed with
+	// Can be accessed without locking only after Complete() returns
```
Remove this sentence. This is not true. This field should never be accessed without locking.

jrajahalme replied:
removed.

rlenglet commented on pkg/completion/completion.go:
```diff
@@ -103,39 +134,48 @@ type Completion struct {
+			c.callback(err)
+		}
+		// Cancel the WaitGroup on failure
+		if err != nil {
```
Just a nitpick:
```
if err != nil && c.cancel != nil {
    c.cancel()
}
```

jrajahalme replied:
fixed. Thanks for the reviews!

rlenglet reviewed (approved):
LGTM! Just two more very minor comments. Feel free to merge as-is, or after you address them.

jrajahalme committed:
xds: Start versioning at 1.

Reserve version 0 for the initial Envoy xDS request (empty
string). This simplifies version number handling as we do not need to
deal with uint64 pointers any more.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Pass error detail when NACK

Define a 'ProxyError' error type that embodies the error detail
received from Envoy when xDS protocols reject resource updates.

This helps debugging when policy or listener updates fail.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Make NACK cancel the WaitGroup

Allow the completion type to complete in failure. The callback is
extended with an error parameter to pass the error condition. The err
parameter will be nil when called after positiva acknowledgement has
been received, non-nil when negative acknowledgement has been recived,
and 'context.Canceled' or 'context.DeadlineExceeded' when the context
has been cancelled or timed out, respectively. A non-nil error
parameter causes cancellation of the whole WaitGroup.

WaitGroup.Wait() returns the most severe error value encountered, in
the following order of severity, from highest to lowest:
1. non-context errors
2. context.Canceled
3. context.DeadlineExceeded
4. nil

To correctly detect xDS resource versions that are ACKed or NACKed the
nonce field is changed to be the same as the version in the response
messages. Then, in the following request, the VersionInfo field tells
us the last ACKed version, while the ResponseNonce field tells us the
last processed, but not ACKed version. This means that all versions
after VersionInfo up to and including ResponseNonce are NACKed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
